Add velox parquet scan to test data correctness against presto parquet

qqibrow opened the pull request:
Introduce a standalone Velox Parquet reader program that reads from a Parquet file and prints its content. This program is intended for testing the Velox Parquet reader and comparing the results against Apache-Parquet.

yingsu00 commented on velox/dwio/parquet/tests/reader/ParquetReaderExample.cpp:
```diff
@@ -0,0 +1,72 @@
+  auto rowReader = reader->createRowReader(rowReaderOpts);
+
+  VectorPtr result = BaseVector::create(outputType, 0, pool.get());
+  while (rowReader->next(500, result)) {
```
nit: make 500 a constant variable

yingsu00 commented on velox/dwio/parquet/tests/reader/ParquetReaderExample.cpp:
```diff
@@ -0,0 +1,72 @@
+  VectorPtr result = BaseVector::create(outputType, 0, pool.get());
+  while (rowReader->next(500, result)) {
+    for (vector_size_t i = 0; i < result->size(); i++) {
+      std::cout << result->toString(i) << std::endl;
```
It'll be nice if you can print the first and last few rows of each row group and page. Then make the number to print as a parameter

qqibrow replied:
If a parquet file has 2 row group, each row group has 2 pages. and number to print is 20. then we want to print 20 / 4 = 5 rows in each page? Is the motivation trying to expose more bugs? If so,  printing all should be able to do the same. 
IMO, print first X rows and print all, the 2 modes should be enough for testing at the beginning.

yingsu00 commented on velox/dwio/parquet/tests/reader/ParquetReaderExample.cpp:
```diff
@@ -0,0 +1,72 @@
+            << std::endl;
+
+  auto outputType = reader->rowType();
+  std::cout << "velox type: " << outputType->toString() << std::endl;
```
It would be great to print the encoding type of each page. In the future we can improve the program to specify the options whether to print them or not, and also specify which page we want to print.

qqibrow replied:
Are you looking for output like `parquet-tools`?
```
lniu@devrestricted-lniu:~$ parquet-tools inspect /home/lniu/parquet_test/testArray/test_1.parquet

############ file meta data ############
created_by: parquet-mr-trino version test-version (build n/a)
num_columns: 1
num_rows: 17900
num_row_groups: 1
format_version: 1.0
serialized_size: 285


############ Columns ############
element

############ Column(element) ############
name: element
path: test.list.element
max_definition_level: 3
max_repetition_level: 1
physical_type: INT32
logical_type: Int(bitWidth=32, isSigned=true)
converted_type (legacy): INT_32
compression: GZIP (space_saved: 48%)
```
Yeah it will be useful to show those info, but the short term goal for this program is to use it to test velox parquet reader. I think we can open another issue as a stretch goal.

yingsu00 replied:
> Are you looking for output like parquet-tools?

Yes. And we should also try to make the command line arguments similar to it. This is useful since users won't need to install parquet-tools separately anymore.

yingsu00 commented:
@qqibrow Will you be able to add PR and commit messages why you introduced the program? My guess was that you wanted to introduce a utility exec program to print the file contents as a convenience to Velox users. But my guess may be wrong.

qqibrow commented on velox/dwio/parquet/tests/reader/ParquetReaderExample.cpp:
```diff
@@ -0,0 +1,72 @@
+  auto pool = facebook::velox::memory::addDefaultLeafMemoryPool();
+
+  facebook::velox::dwio::common::ReaderOptions readerOpts{pool.get()};
+  auto reader = new ParquetReader(
```
switched to make_unique. BTW, is there a automatic way to catch such issue?

qqibrow commented on velox/dwio/parquet/tests/reader/ParquetReaderExample.cpp:
```diff
@@ -0,0 +1,72 @@
+    std::cerr << "Usage: " << argv[0] << " <parquet file>" << std::endl;
+    return 1;
+  }
+  std::string filePath{argv[1]};
```
no. it doesn't need to be absolute path. This command works:
```
~/velox_scan_parquet parquet_test/testArray/test_1.parquet
```

qqibrow commented on velox/dwio/parquet/tests/reader/ParquetReaderExample.cpp:
```diff
@@ -0,0 +1,72 @@
+  auto rowReader = reader->createRowReader(rowReaderOpts);
+
+  VectorPtr result = BaseVector::create(outputType, 0, pool.get());
+  while (rowReader->next(500, result)) {
```
added glag `rows_to_read`.

yingsu00 commented on velox/dwio/parquet/tests/reader/PrintParquetFileContent.cpp:
```diff
@@ -0,0 +1,98 @@
+  auto rowReader = reader->createRowReader(rowReaderOpts);
+
+  VectorPtr result = BaseVector::create(outputType, 0, pool.get());
+  uint64_t to_read = FLAGS_rows_to_read;
```
Velox naming convention: to_read -> numRowsToRead

yingsu00 commented on velox/dwio/parquet/tests/reader/PrintParquetFileContent.cpp:
```diff
@@ -0,0 +1,98 @@
+  uint64_t to_read = FLAGS_rows_to_read;
+
+  while (to_read > 0) {
+    int readed = rowReader->next(std::min(FLAGS_batch_size, to_read), result);
```
readed -> numRowsRead or numRowsScanned

yingsu00 reviewed (commented):
@qqibrow Are you going to print more metadata info, e.g. #RG, #pages, etc? I think they are quite useful.

qqibrow commented on velox/dwio/parquet/tests/reader/PrintParquetFileContent.cpp:
```diff
@@ -0,0 +1,98 @@
+using namespace facebook::velox::dwio::common;
+using namespace facebook::velox::parquet;
+
+static bool ValidatePath(const char* flagname, const std::string& value) {
```
I turned on the validation like:
```
DEFINE_validator(file_path, &ValidatePath);
```
the flag must be `file_path`. 
the signature is introduced in https://gflags.github.io/gflags/#validate

qqibrow commented:
@yingsu00 Thanks for review! addressed all your comments. 

> @qqibrow Are you going to print more metadata info, e.g. #RG, #pages, etc? I think they are quite useful.

How about adding those feature in follow up PRs? The top priority for now is to use it to test velox paruqet reader.

yingsu00 commented:
@qqibrow Do you still plan to work on this PR? If not shall we close it for now?
